nvflare/fuel/f3/drivers/aio_grpc_driver.py

```diff
@@ -127,16 +152,29 @@ async def generate_output(self):
         self.logger.debug(f"{self}: generate_output in thread {ct.name}")
         try:
             while True:
-                item = await self.oq.get()
+                item = await self.read_oq()
                 yield item
         except Exception as ex:
             if self.closing:
                 self.logger.debug(f"{self}: connection closed by {type(ex)}: {secure_format_exception(ex)}")
             else:
                 self.logger.debug(f"{self}: generate_output exception {type(ex)}: {secure_format_exception(ex)}")
             self.logger.debug(secure_format_traceback())
+            raise StopIteration()
+
+    async def read_oq(self):
+        # self.oq.get() does not return before an item is placed into the queue. This could cause it to wait for
+        # a long time. If the connection is closed during this time, the coro won't be done immediately.
+        # To be able to cancel the queue read operation, we wrap it into a task so that we can cancel it when
+        # closing the connection. Once cancelled, "await task" will finish with asyncio.CancelledError exception.
+        with self.lock:
+            if self.closing:
+                raise asyncio.CancelledError("cancelled read_oq: connection closed")
 
-        self.logger.debug(f"{self}: done generate_output")
+            # wrap the queue read into a task
+            task = asyncio.create_task(self.oq.get())
+            self.read_task = task
+        return await task
```

Comment:
We might need to apply this change to 2.3 as well?

Reply:
Yes. I'll leave this to you. :-)